File: leetcode/0815-bus-routes/solution.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int source, int target) {
        if (source == target) return 0;

        unordered_map<int, vector<int>> stopToRoutes;
        for (int r = 0; r < (int)routes.size(); r++) {
            for (int stop : routes[r]) {
                stopToRoutes[stop].push_back(r);
            }
        }

        unordered_set<int> visitedRoutes;
        unordered_set<int> visitedStops = {source};
        queue<pair<int,int>> q; // routeIdx, buses

        for (int r : stopToRoutes[source]) {
            visitedRoutes.insert(r);
            q.push({r, 1});
        }

        while (!q.empty()) {
            auto [routeIdx, buses] = q.front();
            q.pop();
            for (int stop : routes[routeIdx]) {
                if (stop == target) return buses;
                if (visitedStops.count(stop)) continue;
                visitedStops.insert(stop);
                for (int nextRoute : stopToRoutes[stop]) {
                    if (!visitedRoutes.count(nextRoute)) {
                        visitedRoutes.insert(nextRoute);
                        q.push({nextRoute, buses + 1});
                    }
                }
            }
        }

        return -1;
    }
};

File: leetcode/0815-bus-routes/solution.py
from collections import defaultdict, deque
from typing import List


class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        if source == target:
            return 0

        stop_to_routes = defaultdict(list)
        for route_idx, route in enumerate(routes):
            for stop in route:
                stop_to_routes[stop].append(route_idx)

        visited_routes = set()
        visited_stops = {source}
        queue = deque()

        # seed with every route reachable directly from the source stop
        for route_idx in stop_to_routes[source]:
            visited_routes.add(route_idx)
            queue.append((route_idx, 1))

        while queue:
            route_idx, buses = queue.popleft()
            for stop in routes[route_idx]:
                if stop == target:
                    return buses
                if stop in visited_stops:
                    continue
                visited_stops.add(stop)
                for next_route in stop_to_routes[stop]:
                    if next_route not in visited_routes:
                        visited_routes.add(next_route)
                        queue.append((next_route, buses + 1))

        return -1
